app/src/main/java/com/zeapo/pwdstore/Application.kt

```diff
@@ -4,52 +4,116 @@
  */
 package com.zeapo.pwdstore
 
+import android.content.Intent
 import android.content.SharedPreferences
+import android.os.Handler
 import androidx.appcompat.app.AppCompatDelegate
 import androidx.appcompat.app.AppCompatDelegate.MODE_NIGHT_AUTO_BATTERY
 import androidx.appcompat.app.AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM
 import androidx.appcompat.app.AppCompatDelegate.MODE_NIGHT_NO
 import androidx.appcompat.app.AppCompatDelegate.MODE_NIGHT_YES
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.LifecycleObserver
+import androidx.lifecycle.OnLifecycleEvent
+import androidx.lifecycle.ProcessLifecycleOwner
 import androidx.preference.PreferenceManager
 import com.github.ajalt.timberkt.Timber.DebugTree
+import com.github.ajalt.timberkt.Timber.d
 import com.github.ajalt.timberkt.Timber.plant
 import com.zeapo.pwdstore.git.config.setUpBouncyCastleForSshj
 import com.zeapo.pwdstore.utils.PreferenceKeys
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import kotlin.time.ExperimentalTime
+import kotlin.time.seconds
 
 @Suppress("Unused")
-class Application : android.app.Application(), SharedPreferences.OnSharedPreferenceChangeListener {
+class Application : android.app.Application(), SharedPreferences.OnSharedPreferenceChangeListener, LifecycleObserver {
 
-    private var prefs: SharedPreferences? = null
+    private lateinit var prefs: SharedPreferences
+    private val coroutineScope = CoroutineScope(Dispatchers.IO)
+    private var authTimeout = 15
+    private var job: Job? = null
+    @Volatile var isAuthEnabled = false
+    @Volatile var requiresAuthentication = true
 
     override fun onCreate() {
         super.onCreate()
         prefs = PreferenceManager.getDefaultSharedPreferences(this)
-        if (BuildConfig.ENABLE_DEBUG_FEATURES || prefs?.getBoolean(PreferenceKeys.ENABLE_DEBUG_LOGGING, false) ==
-            true) {
+        if (BuildConfig.ENABLE_DEBUG_FEATURES || prefs.getBoolean(PreferenceKeys.ENABLE_DEBUG_LOGGING, false)) {
             plant(DebugTree())
         }
-        prefs?.registerOnSharedPreferenceChangeListener(this)
+        prefs.registerOnSharedPreferenceChangeListener(this)
+        isAuthEnabled = prefs.getBoolean(PreferenceKeys.BIOMETRIC_AUTH, false)
+        if (isAuthEnabled) updateAuthTimeout()
         setNightMode()
         setUpBouncyCastleForSshj()
+        ProcessLifecycleOwner.get().lifecycle.addObserver(this)
```

Comment:
Does this observer also need to be removed?

Reply:
Don't think so, Application never calls `onDestroy` so we never know when to remove the observer. None of the articles that I referenced removed the observer.